File: day8/solve.ts
import { solve } from "../runner/typescript";
import { max, sum, groupBy, range } from "lodash";

function parser(input: string) {
  const lines = input.split("\n");
  const size = lines.length;
  const cells = lines
    .flatMap((l, y) =>
      l.split("").map((char, x) => (char == "." ? null : { x, y, char }))
    )
    .filter((c) => !!c);
  const antennae = groupBy(cells, "char");

  // const firstLine = lines[0];
  // const first = lineParser(firstLine);

  return { antennae, size };
}
type Parsed = ReturnType<typeof parser>;

function print(
  size: number,
  antennae: Record<string, { x: number; y: number }[]>,
  antinodes: Set<string>
) {
  const grid = range(size).map(() => range(size).map(() => "."));
  for (const [antenna, locations] of Object.entries(antennae)) {
    for (const { x, y } of locations) {
      grid[y][x] = antenna;
    }
  }
  for (const a of antinodes) {
    const [x, y] = a.split(",").map(Number);
    grid[y][x] = "#";
  }
  console.log(grid.map((r) => r.join("")).join("\n"));
}

function part1({ size, antennae }: Parsed): number {
  const antinodes = new Set<string>();
  for (const x of range(size)) {
    for (const y of range(size)) {
      for (const locations of Object.values(antennae)) {
        for (const { x: ax, y: ay } of locations) {
          for (const { x: bx, y: by } of locations) {
            if (ax == bx && ay == by) continue;
            const adx = ax - x;
            const ady = ay - y;
            const bdx = bx - x;
            const bdy = by - y;
            if (
              (adx == 2 * bdx && ady == 2 * bdy) ||
              (adx == -2 * bdx && ady == -2 * bdy)
            ) {
              antinodes.add(`${x},${y}`);
            }
          }
        }
      }
    }
  }

  return antinodes.size;
}

function part2(values: Parsed): number {
  return 0;
}

solve({
  parser,
  part1,
  // part2,

  part1Tests: [
    [
      "............\n........0...\n.....0......\n.......0....\n....0.......\n......A.....\n............\n............\n........A...\n.........A..\n............\n............",
      14,
    ],
  ],
  part2Tests: [
    // ["aaa", 0],
    // ["a", 0],
  ],
});

File: day8/solve.py
from typing import Callable
from runner.python import solve

Point = tuple[int, int]


def parser(input):
  lines = input.split("\n")
  antennae = [
    [
      (x, y)
      for y, line in enumerate(lines)
      for x, char in enumerate(line)
      if char == antenna_type
    ]
    for antenna_type in {
      char for line in lines for char in line if char != "."
    }
  ]

  return (antennae, len(lines))


def count_antipodes(
  antennae: list[list[Point]],
  generate_antipodes: Callable[[int, int, int, int], set[Point]],
):
  antipodes = {
    point
    for antenna_group in antennae
    for antennaOne in antenna_group
    for antennaTwo in antenna_group
    if antennaOne != antennaTwo
    for point in generate_antipodes(*antennaOne, *antennaTwo)
  }
  return len(antipodes)


def part1(input):
  antennae, size = parser(input)

  def generate_antipodes(ax, ay, bx, by):
    dx = bx - ax
    dy = by - ay
    return [
      (x, y)
      for x, y in [(bx + dx, by + dy), (ax - dx, ay - dy)]
      if 0 <= x < size and 0 <= y < size
    ]

  return count_antipodes(antennae, generate_antipodes)


def part2(input):
  antennae, size = parser(input)

  def generate_antipodes(ax, ay, bx, by):
    dx = bx - ax
    dy = by - ay
    return {
      (ax + dx * i, ay + dy * i)
      for i in range(-size, size)
      if 0 <= ax + dx * i < size and 0 <= ay + dy * i < size
    }

  return count_antipodes(antennae, generate_antipodes)


solve(
  part1,
  "............\n........0...\n.....0......\n.......0....\n....0.......\n......A.....\n............\n............\n........A...\n.........A..\n............\n............",
  14,
  part2,
  "............\n........0...\n.....0......\n.......0....\n....0.......\n......A.....\n............\n............\n........A...\n.........A..\n............\n............",
  34,
)
